cli/src/pcluster/models/common.py

```diff
@@ -239,3 +252,413 @@ def dump_json(self):
+            path = os.path.join(resource_dir, res)
+            if os.path.isdir(path):
+                AWSApi.instance().s3.upload_fileobj(
+                    file_obj=zip_dir(os.path.join(resource_dir, res)),
+                    bucket_name=self.name,
+                    key=self.get_object_key(S3File.CUSTOM_RESOURCES.value, self._custom_artifacts_name),
+                )
+            elif os.path.isfile(path):
+                AWSApi.instance().s3.upload_file(
+                    file_path=os.path.join(resource_dir, res),
+                    bucket_name=self.name,
+                    key=self.get_object_key(S3File.CUSTOM_RESOURCES.value, res),
+                )
+
+    def get_config(self, config_name, version_id=None):
+        """Get config file from S3 bucket."""
+        return self._get_file(file_type=S3File.CONFIGS.value, file_name=config_name, version_id=version_id)
+
+    def get_config_url(self, config_name):
+        """Get config file http url from S3 bucket."""
+        return self._get_file_url(file_name=config_name, file_type=S3File.CONFIGS.value)
+
+    def get_cfn_template(self, template_name, version_id=None):
+        """Get cfn template from S3 bucket."""
+        return self._get_file(file_type=S3File.TEMPLATES.value, file_name=template_name, version_id=version_id)
+
+    def get_cfn_template_url(self, template_name):
+        """Get cfn template http url from S3 bucket."""
+        return self._get_file_url(file_type=S3File.TEMPLATES.value, file_name=template_name)
+
+    def get_resource_url(self, resource_name):
+        """Get custom resource http url from S3 bucket."""
+        return self._get_file_url(file_type=S3File.CUSTOM_RESOURCES.value, file_name=resource_name)
+
+    # --------------------------------------- S3 private functions --------------------------------------- #
+
+    def _generate_s3_bucket_hash_suffix(self):
+        """
+        Generate 16 characters hash suffix for ParallelCluster s3 bucket.
+
+        :return: 16 chars string e.g. 2238a84ac8a74529
+        """
+        return hashlib.sha256((self.account_id + self.region).encode()).hexdigest()[0:16]
+
+    def _get_artifact_dir(self):
+        """Get artifact directory in S3 bucket by stack output or stack/image tag."""
+        if self._bucket_type == BucketType.CLUSTER:
+            try:
+                # Get artifact directory by cluster stack output
+                stack_data = AWSApi.instance().cfn.describe_stack(self._stack_name)
+                self.__artifact_directory = next(
+                    (
+                        out["OutputValue"]
+                        for out in stack_data.get("Outputs", [])
+                        if out["OutputKey"] == "ArtifactS3RootDirectory"
+                    ),
+                    None,
+                )
+                return self.__artifact_directory
+            except AWSClientError:
+                LOGGER.error("No artifact dir found in cluster stack output.")
```

Comment:
describe_stack will throw ClientError is stack doesn't exist, describe_images will return an empty list if image doesn't exist.

Reply:
yes but in the error message we write: "No artifact dir found in cluster stack output." while we get there only if the cluster is not found. In case the value is not found in the Outputs the logic returns None